File: mobile/modules/core/android/src/main/java/com/mentra/core/CoreModule.kt
package com.mentra.core

import com.mentra.core.services.NotificationListener
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CoreModule : Module() {
    private val bridge: Bridge by lazy { Bridge.getInstance() }
    private var coreManager: CoreManager? = null

    override fun definition() = ModuleDefinition {
        Name("Core")

        // Define events that can be sent to JavaScript
        Events("CoreMessageEvent", "onChange")

        OnCreate {
            // Initialize Bridge with Android context and event callback
            Bridge.initialize(
                    appContext.reactContext
                            ?: appContext.currentActivity
                                    ?: throw IllegalStateException("No context available")
            ) { eventName, data -> sendEvent(eventName, data) }

            // initialize CoreManager after Bridge is ready
            coreManager = CoreManager.getInstance()
        }

        // MARK: - Display Commands

        AsyncFunction("displayEvent") { params: Map<String, Any> ->
            coreManager?.handle_display_event(params)
        }

        AsyncFunction("displayText") { params: Map<String, Any> ->
            coreManager?.handle_display_text(params)
        }

        // MARK: - Connection Commands

        AsyncFunction("requestStatus") { coreManager?.handle_request_status() }

        AsyncFunction("connectDefault") { coreManager?.handle_connect_default() }

        AsyncFunction("connectByName") { deviceName: String ->
            coreManager?.handle_connect_by_name(deviceName)
        }

        AsyncFunction("connectSimulated") { coreManager?.handle_connect_simulated() }

        AsyncFunction("disconnect") { coreManager?.handle_disconnect() }

        AsyncFunction("forget") { coreManager?.handle_forget() }

        AsyncFunction("findCompatibleDevices") { modelName: String ->
            coreManager?.handle_find_compatible_devices(modelName)
        }

        AsyncFunction("showDashboard") { coreManager?.handle_show_dashboard() }

        // MARK: - WiFi Commands

        AsyncFunction("requestWifiScan") { coreManager?.handle_request_wifi_scan() }

        AsyncFunction("sendWifiCredentials") { ssid: String, password: String ->
            coreManager?.handle_send_wifi_credentials(ssid, password)
        }

        AsyncFunction("setHotspotState") { enabled: Boolean ->
            coreManager?.handle_set_hotspot_state(enabled)
        }

        // MARK: - Gallery Commands

        AsyncFunction("queryGalleryStatus") { coreManager?.handle_query_gallery_status() }

        AsyncFunction("photoRequest") {
                requestId: String,
                appId: String,
                size: String,
                webhookUrl: String,
                authToken: String,
                compress: String ->
            coreManager?.handle_photo_request(requestId, appId, size, webhookUrl, authToken, compress)
        }

        // MARK: - Video Recording Commands

        AsyncFunction("startBufferRecording") { coreManager?.handle_start_buffer_recording() }

        AsyncFunction("stopBufferRecording") { coreManager?.handle_stop_buffer_recording() }

        AsyncFunction("saveBufferVideo") { requestId: String, durationSeconds: Int ->
            coreManager?.handle_save_buffer_video(requestId, durationSeconds)
        }

        AsyncFunction("startVideoRecording") { requestId: String, save: Boolean ->
            coreManager?.handle_start_video_recording(requestId, save)
        }

        AsyncFunction("stopVideoRecording") { requestId: String ->
            coreManager?.handle_stop_video_recording(requestId)
        }

        // MARK: - RTMP Stream Commands

        AsyncFunction("startRtmpStream") { params: Map<String, Any> ->
            coreManager?.handle_send_rtmp_stream_start(params.toMutableMap())
        }

        AsyncFunction("stopRtmpStream") {
            coreManager?.handle_stop_rtmp_stream()
        }

        AsyncFunction("keepRtmpStreamAlive") { params: Map<String, Any> ->
            coreManager?.handle_keep_rtmp_stream_alive(params.toMutableMap())
        }

        // MARK: - Microphone Commands

        AsyncFunction("microphoneStateChange") {
                requiredDataStrings: List<String>,
                bypassVad: Boolean ->
            coreManager?.handle_microphone_state_change(requiredDataStrings, bypassVad)
        }

        AsyncFunction("restartTranscriber") { coreManager?.restartTranscriber() }

        // MARK: - RGB LED Control

        AsyncFunction("rgbLedControl") {
                requestId: String,
                packageName: String?,
                action: String,
                color: String?,
                ontime: Int,
                offtime: Int,
                count: Int ->
            coreManager?.handle_rgb_led_control(requestId, packageName, action,
                    color, ontime, offtime, count)
        }

        // MARK: - Settings Commands

        AsyncFunction("updateSettings") { params: Map<String, Any> ->
            coreManager?.handle_update_settings(params)
        }

        // MARK: - STT Commands

        AsyncFunction("setSttModelDetails") { path: String, languageCode: String ->
            val context =
                    appContext.reactContext
                            ?: appContext.currentActivity
                                    ?: throw IllegalStateException("No context available")
            com.mentra.core.stt.STTTools.setSttModelDetails(context, path, languageCode)
        }

        AsyncFunction("getSttModelPath") { ->
            val context =
                    appContext.reactContext
                            ?: appContext.currentActivity
                                    ?: throw IllegalStateException("No context available")
            com.mentra.core.stt.STTTools.getSttModelPath(context)
        }

        AsyncFunction("checkSttModelAvailable") { ->
            val context =
                    appContext.reactContext
                            ?: appContext.currentActivity
                                    ?: throw IllegalStateException("No context available")
            com.mentra.core.stt.STTTools.checkSTTModelAvailable(context)
        }

        AsyncFunction("validateSttModel") { path: String ->
            com.mentra.core.stt.STTTools.validateSTTModel(path)
        }

        AsyncFunction("extractTarBz2") { sourcePath: String, destinationPath: String ->
            com.mentra.core.stt.STTTools.extractTarBz2(sourcePath, destinationPath)
        }

        // MARK: - Android-specific Commands

        AsyncFunction("getInstalledApps") {
            val context =
                    appContext.reactContext
                            ?: appContext.currentActivity
                                    ?: throw IllegalStateException("No context available")
            NotificationListener.getInstance(context).getInstalledApps()
        }

        AsyncFunction("hasNotificationListenerPermission") {
            val context =
                    appContext.reactContext
                            ?: appContext.currentActivity
                                    ?: throw IllegalStateException("No context available")
            NotificationListener.getInstance(context).hasNotificationListenerPermission()
        }
    }
}

File: mobile/modules/core/ios/CoreModule.swift
import ExpoModulesCore
import Photos

public class CoreModule: Module {
    public func definition() -> ModuleDefinition {
        Name("Core")

        // Define events that can be sent to JavaScript
        Events("CoreMessageEvent", "onChange")

        OnCreate {
            // Initialize Bridge with event callback
            Bridge.initialize { [weak self] eventName, data in
                self?.sendEvent(eventName, data)
            }
        }

        // MARK: - Display Commands

        AsyncFunction("displayEvent") { (params: [String: Any]) in
            await MainActor.run {
                CoreManager.shared.displayEvent(params)
            }
        }

        AsyncFunction("displayText") { (params: [String: Any]) in
            await MainActor.run {
                CoreManager.shared.displayText(params)
            }
        }

        // MARK: - Connection Commands

        AsyncFunction("getStatus") {
            await MainActor.run {
                CoreManager.shared.getStatus()
            }
        }

        AsyncFunction("connectDefault") {
            await MainActor.run {
                CoreManager.shared.connectDefault()
            }
        }

        AsyncFunction("connectByName") { (deviceName: String) in
            await MainActor.run {
                CoreManager.shared.connectByName(deviceName)
            }
        }

        AsyncFunction("connectSimulated") {
            await MainActor.run {
                CoreManager.shared.connectSimulated()
            }
        }

        AsyncFunction("disconnect") {
            await MainActor.run {
                CoreManager.shared.disconnect()
            }
        }

        AsyncFunction("forget") {
            await MainActor.run {
                CoreManager.shared.forget()
            }
        }

        AsyncFunction("findCompatibleDevices") { (modelName: String) in
            await MainActor.run {
                CoreManager.shared.findCompatibleDevices(modelName)
            }
        }

        AsyncFunction("showDashboard") {
            await MainActor.run {
                CoreManager.shared.showDashboard()
            }
        }

        // MARK: - WiFi Commands

        AsyncFunction("requestWifiScan") {
            await MainActor.run {
                CoreManager.shared.requestWifiScan()
            }
        }

        AsyncFunction("sendWifiCredentials") { (ssid: String, password: String) in
            await MainActor.run {
                CoreManager.shared.sendWifiCredentials(ssid, password)
            }
        }

        AsyncFunction("setHotspotState") { (enabled: Bool) in
            await MainActor.run {
                CoreManager.shared.setHotspotState(enabled)
            }
        }

        // MARK: - Gallery Commands

        AsyncFunction("queryGalleryStatus") {
            await MainActor.run {
                CoreManager.shared.queryGalleryStatus()
            }
        }

        AsyncFunction("photoRequest") {
            (
                requestId: String, appId: String, size: String, webhookUrl: String?,
                authToken: String?, compress: String?
            ) in
            await MainActor.run {
                CoreManager.shared.photoRequest(
                    requestId, appId, size, webhookUrl, authToken, compress
                )
            }
        }

        // MARK: - Video Recording Commands

        AsyncFunction("startBufferRecording") {
            await MainActor.run {
                CoreManager.shared.startBufferRecording()
            }
        }

        AsyncFunction("stopBufferRecording") {
            await MainActor.run {
                CoreManager.shared.stopBufferRecording()
            }
        }

        AsyncFunction("saveBufferVideo") { (requestId: String, durationSeconds: Int) in
            await MainActor.run {
                CoreManager.shared.saveBufferVideo(requestId, durationSeconds)
            }
        }

        AsyncFunction("startVideoRecording") { (requestId: String, save: Bool) in
            await MainActor.run {
                CoreManager.shared.startVideoRecording(requestId, save)
            }
        }

        AsyncFunction("stopVideoRecording") { (requestId: String) in
            await MainActor.run {
                CoreManager.shared.stopVideoRecording(requestId)
            }
        }

        // MARK: - RTMP Stream Commands

        AsyncFunction("startRtmpStream") { (params: [String: Any]) in
            await MainActor.run {
                CoreManager.shared.startRtmpStream(params)
            }
        }

        AsyncFunction("stopRtmpStream") {
            await MainActor.run {
                CoreManager.shared.stopRtmpStream()
            }
        }

        AsyncFunction("keepRtmpStreamAlive") { (params: [String: Any]) in
            await MainActor.run {
                CoreManager.shared.keepRtmpStreamAlive(params)
            }
        }

        // MARK: - Meta Streaming Configuration
        
        AsyncFunction("configureMetaStreaming") { (resolution: String, frameRate: Int) in
            CoreManager.shared.handle_configure_meta_streaming(resolution, frameRate)
        }

        // MARK: - Microphone Commands

        AsyncFunction("setMicState") { (sendPcmData: Bool, sendTranscript: Bool, bypassVad: Bool) in
            await MainActor.run {
                CoreManager.shared.setMicState(sendPcmData, sendTranscript, bypassVad)
            }
        }

        AsyncFunction("restartTranscriber") {
            await MainActor.run {
                CoreManager.shared.restartTranscriber()
            }
        }

        // MARK: - RGB LED Control

        AsyncFunction("rgbLedControl") {
            (
                requestId: String, packageName: String?, action: String, color: String?,
                ontime: Int, offtime: Int, count: Int
            ) in
            await MainActor.run {
                CoreManager.shared.rgbLedControl(
                    requestId: requestId,
                    packageName: packageName,
                    action: action,
                    color: color,
                    ontime: ontime,
                    offtime: offtime,
                    count: count
                )
            }
        }

        // MARK: - Settings Commands

        AsyncFunction("updateSettings") { (params: [String: Any]) in
            await MainActor.run {
                CoreManager.shared.updateSettings(params)
            }
        }

        // MARK: - STT Commands

        AsyncFunction("setSttModelDetails") { (path: String, languageCode: String) in
            STTTools.setSttModelDetails(path, languageCode)
        }

        AsyncFunction("getSttModelPath") { () -> String in
            return STTTools.getSttModelPath()
        }

        AsyncFunction("checkSttModelAvailable") { () -> Bool in
            return STTTools.checkSTTModelAvailable()
        }

        AsyncFunction("validateSttModel") { (path: String) -> Bool in
            return STTTools.validateSTTModel(path)
        }

        AsyncFunction("extractTarBz2") { (sourcePath: String, destinationPath: String) -> Bool in
            return STTTools.extractTarBz2(sourcePath: sourcePath, destinationPath: destinationPath)
        }

        // MARK: - Android Stubs

        AsyncFunction("getInstalledApps") { () -> Any in
            return false
        }

        AsyncFunction("hasNotificationListenerPermission") { () -> Any in
            return false
        }

        // Notification management stubs (iOS doesn't support these features)
        Function("setNotificationsEnabled") { (_: Bool) in
            // No-op on iOS
        }

        Function("getNotificationsEnabled") { () -> Bool in
            return false
        }

        Function("setNotificationsBlocklist") { (_: [String]) in
            // No-op on iOS
        }

        Function("getNotificationsBlocklist") { () -> [String] in
            return []
        }

        AsyncFunction("getInstalledAppsForNotifications") { () -> [[String: Any]] in
            return []
        }

        // MARK: - Media Library Commands

        AsyncFunction("saveToGalleryWithDate") {
            (filePath: String, captureTimeMillis: Int64?) -> [String: Any] in
            let fileURL = URL(fileURLWithPath: filePath)

            guard FileManager.default.fileExists(atPath: filePath) else {
                return ["success": false, "error": "File does not exist"]
            }

            var assetIdentifier: String?
            let semaphore = DispatchSemaphore(value: 0)
            var resultError: Error?

            PHPhotoLibrary.shared().performChanges {
                let creationRequest: PHAssetChangeRequest
                let pathExtension = fileURL.pathExtension.lowercased()

                if ["mp4", "mov", "avi", "m4v"].contains(pathExtension) {
                    // Video
                    creationRequest = PHAssetChangeRequest.creationRequestForAssetFromVideo(
                        atFileURL: fileURL)!
                } else {
                    // Photo
                    creationRequest = PHAssetChangeRequest.creationRequestForAssetFromImage(
                        atFileURL: fileURL)!
                }

                // Set the creation date if provided
                if let captureMillis = captureTimeMillis {
                    let captureDate = Date(
                        timeIntervalSince1970: TimeInterval(captureMillis) / 1000.0)
                    creationRequest.creationDate = captureDate
                    Bridge.log("CoreModule: Setting creation date to: \(captureDate)")
                }

                assetIdentifier = creationRequest.placeholderForCreatedAsset?.localIdentifier
            } completionHandler: { _, error in
                resultError = error
                semaphore.signal()
            }

            semaphore.wait()

            if let error = resultError {
                Bridge.log("CoreModule: Error saving to gallery: \(error.localizedDescription)")
                return ["success": false, "error": error.localizedDescription]
            }

            Bridge.log("CoreModule: Successfully saved to gallery with proper creation date")
            return ["success": true, "identifier": assetIdentifier ?? ""]
        }
    }
}
